### ISSUE TITLE
[Question] Predictions affected by numerics or by data?

### ISSUE THREAD

**Issue by pabloprf** [pytorch/botorch#1525]
Case information:
- Problem with 1D output and 5D inputs.
- 60 trained data points, with fixed gaussian noise.
- Data has already been normalized (inputs between [0,1] and outputs standardized).
- Using `FixedNoiseGP` as model, `ExactMarginalLogLikelihood` as mll and `fit_gpytorch_mll` as training method.

Problem found:

Scanning through variable 3 (fixing all others to the last data point) produces predicted values that seem as if they were affected by a numerical instability, as shown in the plot below. Scanning in other variables does not produce such behavior. The optimized raw lengthscales turn out to be `[-0.8570, -0.9472, -9.4033, -0.8491, -0.8841]`, so this may actually mean that the variable-3 behavior may be due to the optimizer finding an unsuitable lengthscale.

<img width="1156" alt="Screenshot 2022-11-26 at 11 02 18 PM" src="https://user-images.githubusercontent.com/20051119/204118676-17f6dbca-bef9-4ec6-88f8-89122f47ab05.png">

Question:s

- Is this due to a numerical instability or the optimized lengthscales?
- Is it possible to fix this somehow with the choice of model/constraints? The data used for this exercise comes from a simulation code, which sometimes produces discontinuous behavior (eigenvalue solver for linear turbulence). If this problem is caused by the data, is there a way to be robust against this?

## To reproduce

```python
import torch,botorch,gpytorch
import matplotlib.pyplot as plt

# DATA

x = torch.tensor(
		[[0.1646, 0.0000, 0.2568, 0.3778, 0.8667],
		[0.2792, 0.0000, 0.4461, 0.6568, 0.6019],
		[0.0549, 0.0000, 0.6456, 0.5522, 0.3870],
		[0.7618, 0.0000, 0.8698, 0.8060, 0.4118],
		[0.5640, 0.0000, 0.3833, 0.9315, 0.2071],
		[0.6265, 0.0000, 0.7617, 0.7944, 0.9364],
		[0.8360, 0.0000, 0.9129, 0.4781, 0.1383],
		[0.3087, 0.0000, 0.1892, 0.2926, 0.5415],
		[0.4979, 0.0000, 0.0603, 0.1964, 0.7682],
		[0.9671, 0.0000, 0.5945, 0.0545, 0.0424],
		[1.0000, 0.9215, 1.0000, 0.1453, 0.1245],
		[0.8095, 0.9126, 0.8165, 0.5558, 0.7507],
		[0.7555, 0.9680, 0.6699, 0.2302, 0.5154],
		[0.6297, 0.9008, 0.5526, 0.0000, 0.3777],
		[0.5311, 0.8666, 0.4589, 0.1651, 0.4956],
		[0.6008, 0.7273, 0.6444, 1.0000, 1.0000],
		[0.3430, 0.7805, 0.0823, 0.6518, 0.8593],
		[0.7432, 0.7624, 0.0823, 0.2231, 0.5869],
		[0.1176, 0.9173, 0.3013, 0.2133, 0.4042],
		[0.0000, 0.8525, 0.0000, 0.8125, 0.8098],
		[0.2577, 0.8148, 0.1838, 0.3284, 0.5352],
		[0.6420, 0.7401, 0.4436, 0.2697, 0.1503],
		[0.4180, 0.7739, 0.2604, 0.0308, 0.0000],
		[0.1618, 0.8544, 0.1717, 0.5200, 0.6328],
		[0.1548, 1.0000, 0.0872, 0.4770, 0.7683],
		[0.3873, 0.8970, 0.4391, 0.4140, 0.6616],
		[0.2620, 0.8879, 0.3085, 0.5410, 0.6371],
		[0.2295, 0.8730, 0.2392, 0.5796, 0.6567],
		[0.2323, 0.8642, 0.2063, 0.6731, 0.6275],
		[0.1940, 0.8769, 0.0587, 0.7057, 0.6647],
		[0.1273, 0.8323, 0.2713, 0.5256, 0.7150],
		[0.1269, 0.9394, 0.1734, 0.8415, 0.6894],
		[0.1952, 0.8186, 0.2004, 0.7875, 0.6542],
		[0.1668, 0.8618, 0.2024, 0.8737, 0.6785],
		[0.2513, 0.8294, 0.1831, 0.6771, 0.6297],
		[0.1710, 0.8790, 0.2234, 0.4661, 0.6648],
		[0.2562, 0.8158, 0.1691, 0.6666, 0.5775],
		[0.2551, 0.8208, 0.1675, 0.6830, 0.6367],
		[0.2250, 0.8428, 0.1715, 0.7536, 0.6387],
		[0.2522, 0.8370, 0.1543, 0.6747, 0.6332],
		[0.2231, 0.8266, 0.1650, 0.7245, 0.6324],
		[0.2393, 0.8324, 0.1653, 0.6983, 0.6244],
		[0.2422, 0.8337, 0.1633, 0.6889, 0.6231],
		[0.2135, 0.8158, 0.1718, 0.7291, 0.6287],
		[0.2075, 0.8217, 0.1600, 0.6923, 0.6481],
		[0.2031, 0.8138, 0.1723, 0.7136, 0.6287],
		[0.2214, 0.8319, 0.1622, 0.7117, 0.6302],
		[0.2143, 0.8392, 0.1740, 0.7848, 0.6525],
		[0.2465, 0.8640, 0.1647, 0.6994, 0.6445],
		[0.1744, 0.8437, 0.1613, 0.8060, 0.6529],
		[0.2165, 0.8451, 0.1688, 0.7682, 0.6504],
		[0.2533, 0.8332, 0.1599, 0.6683, 0.5970],
		[0.2122, 0.8308, 0.1668, 0.7290, 0.6558],
		[0.2185, 0.8243, 0.1637, 0.7201, 0.6360],
		[0.2223, 0.8438, 0.1654, 0.7188, 0.6470],
		[0.2255, 0.8257, 0.1631, 0.7045, 0.6396],
		[0.2139, 0.8210, 0.1653, 0.7137, 0.6373],
		[0.2276, 0.8281, 0.1607, 0.7007, 0.6444],
		[0.2124, 0.8232, 0.1640, 0.7027, 0.6416],
		[0.2095, 0.8383, 0.1630, 0.7556, 0.6403]], dtype=torch.float64)

y = torch.tensor([[-1.3534, -1.3534, -1.3534, -1.3534, -1.3534, -1.3534, -1.3534, -1.3534,
		 -1.3534, -1.3534,  0.5711, -0.3643,  1.2032,  0.5475, -0.3860, -1.5254,
		 -1.2140, -0.9930,  1.6627, -0.7041, -0.1376, -1.4207, -1.2592, -0.7330,
		  0.9924,  0.4128,  0.8549,  1.0173,  1.2040, -0.2757,  0.0075,  0.3151,
		  0.3006,  1.1489,  0.5764,  1.4265, -0.0483,  0.0257,  0.8988, -0.7682,
		  0.5959,  0.7016,  0.7326,  0.3820, -0.9366,  0.3551,  0.7115,  0.8202,
		 -0.5668, -0.5629,  1.2598,  1.0019,  0.6780,  0.5600,  2.2277,  0.5892,
		  0.4993,  1.4503,  0.5475, -0.8490]], dtype=torch.float64).transpose(0,1)

yv = torch.tensor([[8.4447e-03, 8.4447e-03, 8.4447e-03, 8.4447e-03, 8.4447e-03, 8.4447e-03,
		 8.4447e-03, 8.4447e-03, 8.4447e-03, 8.4447e-03, 9.2589e-03, 2.4458e-03,
		 1.6340e-02, 9.0334e-03, 2.3396e-03, 7.3995e-05, 4.8561e-05, 3.2460e-04,
		 2.2741e-02, 1.0538e-03, 3.6952e-03, 1.1323e-05, 2.2156e-05, 9.6209e-04,
		 1.3756e-02, 7.7978e-03, 1.2191e-02, 1.4050e-02, 1.6350e-02, 2.9035e-03,
		 4.6299e-03, 6.9595e-03, 6.8392e-03, 1.5654e-02, 9.3097e-03, 1.9319e-02,
		 4.2576e-03, 4.7544e-03, 1.2680e-02, 8.5607e-04, 9.4990e-03, 1.0557e-02,
		 1.0878e-02, 7.5288e-03, 4.3431e-04, 7.2974e-03, 1.0659e-02, 1.1811e-02,
		 1.5465e-03, 1.5622e-03, 1.7071e-02, 1.3867e-02, 1.0316e-02, 9.1521e-03,
		 3.2060e-02, 9.4334e-03, 8.5810e-03, 1.9651e-02, 9.0329e-03, 6.3594e-04]],
	   dtype=torch.float64).transpose(0,1)

# MODEL

model =botorch.models.gp_regression.FixedNoiseGP(x,y,yv)
mll = gpytorch.mlls.ExactMarginalLogLikelihood( model.likelihood, model )
model.train();  model.likelihood.train();   mll.train()

# TRAINING

mll = botorch.fit.fit_gpytorch_mll(	mll, optimizer_kwargs = {'options': {'disp':True}} )	
for param_name, param in model.named_parameters(): print(f'{param_name} = {param}')

# PLOTTING

num_points_scan = 100000

plt.ion(); fig, axs = plt.subplots(nrows=2,figsize=(14,8))
for cont,i in enumerate([1,2]):

	x_test = torch.repeat_interleave(x[-1,:].unsqueeze(0),num_points_scan,dim=0)
	x_test_var = torch.linspace(0,1,num_points_scan)
	x_test[:,i] = x_test_var

	with torch.no_grad(), gpytorch.settings.eval_cg_tolerance(1E-6):
		posterior    = model.posterior(x_test)
		mean      	 = posterior.mean
		lower, upper = posterior.mvn.confidence_region()

	ax = axs[cont]
	ax.plot(x_test_var.detach().numpy(),mean[:,0].detach().numpy(),'-o',lw=0.5,markersize=0.5)
	ax.fill_between(x_test_var.detach().numpy(), lower.detach().numpy(), upper.detach().numpy(), alpha=0.3)

	ax.set_ylabel('y')
	ax.set_xlabel(f'Scan of variable {i}')

```

**Comment by Balandat** [comment]
> Is this due to a numerical instability or the optimized lengthscales?

I would suspect that the behavior you're seeing is largely due to the tiny lengthscale associated with the third variable. This can in turn cause numerical issues, but these are more a symptom than they are the cause. 

> The data used for this exercise comes from a simulation code, which sometimes produces discontinuous behavior (eigenvalue solver for linear turbulence). If this problem is caused by the data, is there a way to be robust against this?

So if the DGP can indeed be discontinuous then this is going to be a challenge for a GP model, which by definition assumes continuity (smoothness, in fact). Often what people do is that they infer a noise level (possibly in addition to the provided ones) which can help smooth out things near an actual discontinuity, but of course that is going to cost you accuracy if you actually want to model this discontinuity. If the noise level is small and there is a lot of data in the neighborhood of a discontinuity then the only way for the model to explain this data as samples from a MVN is by really cranking down the inferred lengthscales.

If capturing this discontinuity is important, then a GP might not be the best model. There is also work on so-called change point detection models - this essentially amounts to learning the location of the discontinuity (ties), and the stitching together multiple GP models around that.  

If capturing the discontinuity is not very important for your purposes then maybe something can be done by some data preprocessing, i.e. removing some data in the neighborhood of the discontinuity or fitting separate models (this would be akin to the change point detection models if you already know where the discontinuity is).

**Comment by pabloprf** [comment]
Thanks @Balandat for your suggestions.
It is not important to capture the discontinuity for my case. In fact, the simulation code I'm using to produce the data is already a reduced version of a more complex and expensive code, and the appearance of such discontinuous behavior is not physical, but rather due to the numerical approaches in the reduced model. From your answer it sounds like the approach of inferring a noise level above the supplied one may give be better results. At some point I would like to implement a non-stationary kernel or a change point detection model, because the physical process I'm trying to model with the GPs has a so-called "critical gradient" behavior, which means that there's a change in slope that can be rather dramatic. In such a case, that would be real physics to capture.

**Comment by Balandat** [comment]
> From your answer it sounds like the approach of inferring a noise level above the supplied one may give be better results

Yeah that would at least be very easy to try. GPyTorch actually offers that ability in the `FixedNoiseGaussianLikelihood` that is used in the `FixedNoiseGP` in botorch: https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/likelihoods/gaussian_likelihood.py#L203

I don't think this is easily exposed as of now, but you should be able to either edit the code locally or just make a subclass that redefines the constructor and passes the relevant arg here :https://github.com/pytorch/botorch/blob/main/botorch/models/gp_regression.py#L256-L258

Curious to see how this goes and whether this will work well for your dataset, please keep me posted.
